common/network-types/src/session/protocol.rs

```diff
@@ -0,0 +1,403 @@
+            SessionMessage::Acknowledge(a) => write!(f, "acknowledgement of {:?}", a.0),
+        }
+    }
+}
+
+impl<const C: usize> SessionMessage<C> {
+    /// Header size of the session message.
+    /// This is currently the version byte and the size of [SessionMessageDiscriminants] representation.
+    pub const HEADER_SIZE: usize = 1 + mem::size_of::<SessionMessageDiscriminants>();
+
+    /// Current version of the protocol.
+    pub const VERSION: u8 = 0;
+
+    /// Maximum number of segments per frame.
+    pub const MAX_SEGMENTS_PER_FRAME: usize = SegmentRequest::<C>::MAX_MISSING_SEGMENTS_PER_FRAME;
+
+    /// Returns the minimum size of a [SessionMessage].
+    pub fn minimum_message_size() -> usize {
+        // Make this a "const fn" once "min" is const fn too
+        (Self::HEADER_SIZE + Segment::HEADER_SIZE + 1)
+            .min(SegmentRequest::<C>::SIZE)
+            .min(FrameAcknowledgements::<C>::SIZE)
+    }
+
+    /// Convenience method to encode the session message.
+    pub fn into_encoded(self) -> Box<[u8]> {
+        Vec::from(self).into_boxed_slice()
+    }
+}
+
+impl<const C: usize> TryFrom<&[u8]> for SessionMessage<C> {
+    type Error = SessionError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() < Self::minimum_message_size() || value.len() > C {
+            return Err(SessionError::IncorrectMessageLength);
+        }
+
+        let version = value[0];
+        if version != Self::VERSION {
+            return Err(SessionError::WrongVersion);
+        }
+
+        match SessionMessageDiscriminants::from_repr(value[1]).ok_or(SessionError::UnknownMessageTag)? {
+            SessionMessageDiscriminants::Segment => Ok(SessionMessage::Segment(
+                (&value[2..]).try_into().map_err(|_| SessionError::ParseError)?,
+            )),
+            SessionMessageDiscriminants::Request => Ok(SessionMessage::Request((&value[2..]).try_into()?)),
+            SessionMessageDiscriminants::Acknowledge => Ok(SessionMessage::Acknowledge((&value[2..]).try_into()?)),
+        }
+    }
+}
+
+impl<const C: usize> From<SessionMessage<C>> for Vec<u8> {
+    fn from(value: SessionMessage<C>) -> Self {
+        let mut ret = Vec::with_capacity(C);
+        ret.push(SessionMessage::<C>::VERSION);
+        ret.push(SessionMessageDiscriminants::from(&value) as u8);
+
+        match value {
+            SessionMessage::Segment(s) => ret.extend(Vec::from(s)),
+            SessionMessage::Request(b) => ret.extend(Vec::from(b)),
+            SessionMessage::Acknowledge(f) => ret.extend(Vec::from(f)),
+        };
+
+        ret.shrink_to_fit();
+        ret
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::session::Frame;
+    use bitvec::bitarr;
+    use hex_literal::hex;
+    use hopr_platform::time::native::current_time;
+    use rand::{thread_rng, Rng};
+    use std::time::SystemTime;
+
+    #[test]
+    fn test_session_message_segment() -> anyhow::Result<()> {
```

Comment:
I more recently realized that tests are more readable, if the `test_<component>` is just dropped from the name.

```suggestion
    fn session_message_is_reconstructable_from_the_segments
```